wip

jamestrew opened the pull request:
# Description

Fixes #2272 

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list relevant details about your configuration

- [ ] Test A
- [ ] Test B

**Configuration**:
* Neovim version (nvim --version):
* Operating system and version:

# Checklist:

- [ ] My code follows the style guidelines of this project (stylua)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (lua annotations)

jamestrew commented:
Maybe worth discussing is that with the currently implementation of the PR, for `grep_string` the rg submatch highlights aren't applied to de-conflict with `grep_string`'s sorter matches.

![image](https://github.com/nvim-telescope/telescope.nvim/assets/66286082/600546e8-8a0a-4028-b8e9-cd03ae23ecf3)

Maybe it's better to use a different highlight for the rg submatches (not `TelescopeMatching`) for both pickers. 
![image](https://github.com/nvim-telescope/telescope.nvim/assets/66286082/2b09cf4c-28e6-49a3-8fc5-58ae15f259e3)
^ here notices the `bar` submatches are probably highlighted in red while the `grep_string` prompt matches are still highlighted in the cyan (TelescopeMatching).

jamestrew commented:
Yeah I'm not opposed to keeping `gen_from_vimgrep` and calling this new entry_maker something else for backwards compatibility reasons.

> Another question i need to think about is if displayer is the best way to do highlighting here.

Yeah I don't love creating a new `displayer` for each entry. I'm open to ideas.

Marskey commented:
when this will be merged into master, I realy love this new feature

Marskey commented:
"<cmd> Telescope pickers <CR> will occur error:
![image](https://github.com/nvim-telescope/telescope.nvim/assets/13499986/d54566fa-f47b-4d77-89a5-917ad62a00de)

Conni2461 commented:
> Yeah I don't love creating a new displayer for each entry. I'm open to ideas.

Displayer is just a wrapper around the actual interface, i'll going to rewrite that section.

I am also making this PR a Draft for the time being. I know its almost done but then we might have less ppl asking :laughing:

jamestrew commented:
Sounds good.

> " Telescope pickers will occur error: ![image](https://user-images.githubusercontent.com/13499986/254756434-d54566fa-f47b-4d77-89a5-917ad62a00de.png)

I also gotta fix this eventually as well

jamestrew commented:
Note to self: vim.base64?

Marskey commented:
> " Telescope pickers will occur error: ![image](https://private-user-images.githubusercontent.com/13499986/254756434-d54566fa-f47b-4d77-89a5-917ad62a00de.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY3ODEzNjcsIm5iZiI6MTcwNjc4MTA2NywicGF0aCI6Ii8xMzQ5OTk4Ni8yNTQ3NTY0MzQtZDU0NTY2ZmEtZjQ3Yi00ZDc3LTg5YTUtOTE3YWQ2MmEwMGRlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAxVDA5NTEwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUyMzU4OTNjMTdiNzU0YWM1ODM1NWU5YWFkODg0ODYyY2VmMmVlMGM3NWQ4NzJhMzNmYjUxNjIyMzJhOGI1OTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1zqYXSaeuKFWG2abWPvJJ80YE55YQgPeZ59iCnRR9wE)

```
function make_entry.gen_from_vimgrep_json(opts)
  opts = opts or {}
  local cwd = vim.fn.expand(opts.cwd or vim.loop.cwd() or "")

  local function bytes_or_text_to_str(data)
    if data.bytes ~= nil then
      return require("telescope.algos.base64").decode(data.bytes)
    else
      return data.text
    end
  end

  local mt_vimgrep_entry = {}
  mt_vimgrep_entry.display = function(entry)
    local display_filename = utils.transform_path(opts, entry.filename)

    local coordinates = ":"
    if not opts.disable_coordinates then
      if entry.lnum then
        if entry.col then
          coordinates = string.format(":%s:%s:", entry.lnum, entry.col)
        else
          coordinates = string.format(":%s:", entry.lnum)
        end
      end
    end

    local file_pos, hl_group, icon = utils.transform_devicons(
      entry.filename,
      string.format("%s%s", display_filename, coordinates),
      opts.disable_devicons
    )

    return file_pos .. entry.text,
      (function()
        if opts.__finder == "grep_string" then
          return {}
        end

        local highlights = { { { 0, #icon }, hl_group or "" } }
        local file_pos_len = #file_pos
        for _, submatch in ipairs(entry.submatches) do
          table.insert(
            highlights,
            { { submatch["start"] + file_pos_len, submatch["end"] + file_pos_len }, "TelescopeMatching" }
          )
        end
        return highlights
      end)()
  end

  mt_vimgrep_entry.__index = function(t, k)
    local override = handle_entry_index(opts, t, k)
    if override then
      return override
    end

    local raw = rawget(mt_vimgrep_entry, k)
    if raw then
      return raw
    end

    if k == "path" then
      local path = Path:new(t.filename)
      if path:is_absolute() then
        return t.filename
      end
      return Path:new({ cwd, t.filename }):absolute()
    end

    if k == "text" then
      return t.value
    end

    if k == "ordinal" then
      local text = t.text
      return opts.only_sort_text and text or text .. " " .. t.filename
    end
  end

  return function(line)
    local msg = vim.json.decode(line)
    if msg == nil or msg.type ~= "match" then
      return
    end

    local text = bytes_or_text_to_str(msg.data.lines):gsub("%s+$", "")
    return setmetatable({
      value = text,
      filename = bytes_or_text_to_str(msg.data.path),
      lnum = msg.data.line_number,
      col = #msg.data.submatches ~= 0 and msg.data.submatches[1].start + 1 or nil,
      submatches = msg.data.submatches,
    }, mt_vimgrep_entry)
  end
end
```
@jamestrew  this should fix it

jamestrew commented:
@Marskey thanks I've incorporated your snippets. looks to be working well so far.

Marskey commented:
```lua
    return setmetatable({
      value = text,
      filename = bytes_or_text_to_str(msg.data.path),
      lnum = msg.data.line_number,
      col = #msg.data.submatches ~= 0 and msg.data.submatches[1].start + 1 or nil,
      colend = #msg.data.submatches ~= 0 and msg.data.submatches[1]["end"] + 1 or nil,
      submatches = msg.data.submatches,
    }, mt_vimgrep_entry)

```
"Please consider whether `colend` needs to be included."

jamestrew commented:
I also rebased everything.
I switched out to using `vim.base64` over some copy/pasted code so this is now requires neovim 0.10. I think the intention was to make telescope 0.2.0 require nvim 0.10 so I'm leaning on keeping this.

humblehacker commented:
This fix has been working for me. Any chance we can get this merged?

Marskey commented:
til neovim 0.10 released, or you can merged by youself if you are using pre-release version 0.10

npearson72 commented:
Neovim 0.10 has already been released. Any reason this PR still can't be merged?

Conni2461 commented:
I still need to review this, thats missing.

jamestrew commented:
We're still supporting 0.9 on the master branch as well.
